runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/grpcclient/GrpcGetWorkerMetadataStreamTest.java

```diff
@@ -0,0 +1,338 @@
+      Lists.newArrayList(
+          WorkerMetadataResponse.Endpoint.newBuilder().setDirectEndpoint("direct_path").build());
+  private static final Map<String, WorkerMetadataResponse.Endpoint> GLOBAL_DATA_ENDPOINTS =
+      Maps.newHashMap();
+  private static final JobHeader TEST_JOB_HEADER =
+      JobHeader.newBuilder()
+          .setJobId("test_job")
+          .setWorkerId("test_worker")
+          .setProjectId("test_project")
+          .build();
+  private static final String FAKE_SERVER_NAME = "Fake server for GrpcGetWorkerMetadataStreamTest";
+  @Rule public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();
+  private final MutableHandlerRegistry serviceRegistry = new MutableHandlerRegistry();
+  private final Set<AbstractWindmillStream<?, ?>> streamRegistry = new HashSet<>();
+  private ManagedChannel inProcessChannel;
+  private GrpcGetWorkerMetadataStream stream;
+
+  private GrpcGetWorkerMetadataStream getWorkerMetadataTestStream(
+      GetWorkerMetadataTestStub getWorkerMetadataTestStub,
+      AtomicLong metadataVersion,
+      Consumer<WindmillEndpoints> endpointsConsumer) {
+    serviceRegistry.addService(getWorkerMetadataTestStub);
+    return GrpcGetWorkerMetadataStream.forTesting(
+        responseObserver ->
+            CloudWindmillServiceV1Alpha1Grpc.newStub(inProcessChannel)
+                .getWorkerMetadataStream(responseObserver),
+        metadataVersion,
+        TEST_JOB_HEADER,
+        streamRegistry,
+        endpointsConsumer);
+  }
+
+  @Before
+  public void setUp() throws IOException {
+    Server server =
+        InProcessServerBuilder.forName(FAKE_SERVER_NAME)
+            .fallbackHandlerRegistry(serviceRegistry)
+            .directExecutor()
+            .build()
+            .start();
+    inProcessChannel =
+        grpcCleanup.register(
+            InProcessChannelBuilder.forName(FAKE_SERVER_NAME).directExecutor().build());
+    grpcCleanup.register(server);
+    grpcCleanup.register(inProcessChannel);
+    GLOBAL_DATA_ENDPOINTS.put(
+        "global_data",
+        WorkerMetadataResponse.Endpoint.newBuilder().setDirectEndpoint("global_data").build());
+  }
+
+  @After
+  public void cleanUp() {
+    inProcessChannel.shutdownNow();
+  }
+
+  @Test
+  public void testGetWorkerMetadata() {
+    WorkerMetadataResponse mockResponse =
+        WorkerMetadataResponse.newBuilder()
+            .setMetadataVersion(1)
+            .addAllDirectPathEndpoints(DIRECT_PATH_ENDPOINTS)
+            .putAllGlobalDataEndpoints(GLOBAL_DATA_ENDPOINTS)
+            .build();
+    TestWindmillEndpointsConsumer testWindmillEndpointsConsumer =
+        new TestWindmillEndpointsConsumer();
+    TestGetWorkMetadataRequestObserver requestObserver =
+        new TestGetWorkMetadataRequestObserver(mockResponse);
+    GetWorkerMetadataTestStub testStub = new GetWorkerMetadataTestStub(requestObserver);
+    AtomicLong metadataVersion = new AtomicLong(0);
```

Comment:
nit: use some non-0 version, 0 might be correct if just default initialized on accident instead of propagated for example.

(sorry if some of these are duplicated, I was putting comments on some other snapshot I think, not sure where they'll show up).

Reply:
done. no worries!